File: heaps/GroupAnagrams.py
#---------------------------------------------------------------
# Solution 1
#---------------------------------------------------------------
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # sort and group
        dict = collections.defaultdict(list)
        # return values
        for s in strs:
            sorted_list = list(s)
            sorted_list.sort()
            sorted_string = ''.join(sorted_list)        
            dict[sorted_string].append(s)
        return dict.values()
#---------------------------------------------------------------
# Solution 2
#---------------------------------------------------------------
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        groups = collections.defaultdict(list)
        for s in strs:
             groups[tuple(sorted(s))].append(s)
        return map(sorted, groups.values())

#----------------------------------------------------------------
# Solution 3
#----------------------------------------------------------------
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # sort and group
        groups = itertools.groupby(sorted(strs, key=sorted), sorted)
        # return values
        return [sorted(members) for _, members in groups]

File: heaps/GroupAnagrams.cs
public class Solution {
    public IList<IList<string>> GroupAnagrams(string[] strs) {
        // give that char is an unsigned 16-bit value.
        // algorithm: sort letters and compare using a dictionary.
        // return the dictionary values
        Dictionary<string, IList<string>> dict = new Dictionary<string, IList<string>>();
        foreach(string str in strs){
            // Sort Array and Return string is O(nlogn + 2*n) = O(1)
            string sorted = new string (str.OrderBy(c => c).ToArray());
            if(dict.ContainsKey(sorted)){
                dict[sorted].Add(str);
            } else {
                dict[sorted] = new List<string>{str};
            }
        }
        return dict.Values.ToList();
    }
}